Sources/SourceKitLSP/Swift/OnDiskDocumentManager.swift

```diff
@@ -0,0 +1,169 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2025 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import BuildServerIntegration
+import Foundation
+import IndexStoreDB
+import LanguageServerProtocol
+import SKLogging
+import SKUtilities
+import SwiftExtensions
+
+/// Caches document snapshots from disk opened in sourcekitd.
+///
+/// Used by `textDocument/doccDocumentation` requests to retrieve symbol graphs for files that are not currently
+/// open in the editor. This allows for retrieving multiple symbol graphs from the same file without having
+/// to re-open and parse the syntax tree every time.
+actor OnDiskDocumentManager: Sendable {
+  private weak var sourceKitLSPServer: SourceKitLSPServer?
+  private var openSnapshots: [DocumentURI: (snapshot: DocumentSnapshot, patchedCompileCommand: SwiftCompileCommand?)]
+
+  init(sourceKitLSPServer: SourceKitLSPServer) {
```

Comment:
Should this be `fileprivate` to ensure that the `OnDiskDocumentManager` is always created using `withOnDiskDocumentManager`?